dask_cuda/explicit_comms/dataframe_shuffle.py

```diff
@@ -0,0 +1,260 @@
+    else:
+        return concat(df_parts)
+
+
+def partition_by_hash(df, columns, n_chunks, ignore_index=False):
+    """Splits dataframe into partitions
+
+    The partitions is determined by the hash value of the rows in `columns`.
+
+    Parameters
+    ----------
+    df: DataFrame
+    columns: label or list
+        Column names on which to split the dataframe
+    npartition: int
+        Number of partitions
+    ignore_index : bool, default False
+        Set True to ignore the index of `df`
+
+    Returns
+    -------
+    out: Dict[int, DataFrame]
+        A dictionary mapping integers in {0..npartition} to dataframes.
+    """
+    if df is None:
+        return [None] * n_chunks
+
+    # Hashing `columns` in `df` and assing it to the "_partitions" column
+    df["_partitions"] = partitioning_index(df[columns], n_chunks)
+    # Split `df` based on the hash values in the "_partitions" column
+    try:
+        # For Dask < 2.17 compatibility
+        ret = shuffle_group(df, "_partitions", 0, n_chunks, n_chunks, ignore_index)
+    except TypeError:
+        ret = shuffle_group(
+            df, "_partitions", 0, n_chunks, n_chunks, ignore_index, n_chunks
+        )
+
+    # Let's remove the partition column and return the partitions
+    del df["_partitions"]
+    for df in ret.values():
+        del df["_partitions"]
+    return ret
+
+
+async def rearrange_by_column(n_chunks, rank, eps, left_table, column):
+    left_bins = partition_by_hash(left_table, column, n_chunks, ignore_index=True)
+    return await exchange_and_concat_bins(rank, eps, left_bins)
+
+
+async def _rearrange_by_column(s, workers, dfs_nparts, dfs_parts, column):
+    """
+    Parameters
+    ----------
+    s: dict
+        Worker session state
+    workers: set
+        Set of ranks of all the participants
+    dfs_nparts: list of dict
+        List of dict that for each worker rank specifices the
+        number of partitions that worker has. If the worker doesn't
+        have any partitions, it is excluded from the dict.
+        E.g. `dfs_nparts[0][1]` is how many partitions of the "left"
+        dataframe worker 1 has.
+    dfs_parts: list of lists of Dataframes
+        List of inputs, which in this case are two dataframe lists.
+    column : label or list, or array-like
+        The bases of the rearrangement.
+    """
+    assert s["rank"] in workers
+    df_parts = dfs_parts[0]
+
+    # Trimming such that all participanting workers get a rank within 0..len(workers)
+    trim_map = {}
+    for i in range(s["nworkers"]):
+        if i in workers:
+            trim_map[i] = len(trim_map)
+
+    rank = trim_map[s["rank"]]
+    eps = {trim_map[i]: s["eps"][trim_map[i]] for i in workers if i != s["rank"]}
+
+    df = df_concat(df_parts)
+
+    return await rearrange_by_column(len(workers), rank, eps, df, column)
+
+
+def dataframe_rearrange_by_column(df, column):
+    """Order divisions of DataFrame so that all values within column(s) align
+
```

Comment:
Perhaps its best to state that this is an "explicit-comms" version of `rearrange_by_column_tasks`?